Switch to LLVM/clang 16 (16.0.3)

azat opened the pull request:
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Switch to LLVM/clang 16

Based on: #47709 (cc @alexey-milovidov )
Reverts: #43195
Fixes: #41500 (many thanks to @itrofimow for the patch!)
Closes: #46383 (cc @Felixoid )
Partially reverts: #42457/#40873 (workaround for emitting .debug_aranges), since [patch](https://reviews.llvm.org/D133092) had been merged to upstream) and included into LLVM 16
Blocks: #47783

Required fixes:
- https://github.com/VectorCamp/vectorscan/pull/148
- https://github.com/VectorCamp/vectorscan/pull/149
- https://github.com/ClickHouse/woboq_codebrowser/pull/8
- https://github.com/ClickHouse/libfiu/pull/2

alexey-milovidov committed:
Use LLVM 16

alexey-milovidov committed:
Update one more place

azat committed:
Merge remote-tracking branch 'u/master' into build/llvm-16

azat committed:
Enable back thread fuzzuer under TSan

Since initial issue got resolved [1].

  [1]: https://github.com/google/sanitizers/issues/1540

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Disable .debug_aranges workaround under ThinLTO for clang/llvm 16

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Revert "Merge pull request #43195 from ClickHouse/workaround-for-llvm-bug"

With clang 16 this workaround should not be required anymore.

This reverts commit 992729f3c0cb00611f0e2c082cf996d6ea718a22, reversing
changes made to d885b3cc4c5514e4e8a090e085041e589060a03a.

robot-ch-test-poll commented:
<!-- automatic status comment for PR #49678 from azat/ClickHouse:build/llvm-16 -->
This is an automated comment for commit 23569c95e3be55d4720ffd95c1c5c658539a60fe with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49678/23569c95e3be55d4720ffd95c1c5c658539a60fe/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

azat committed:
docker: install libclang-16-dev for woboq code browser

Otherwise you will get the following error:

    CMake Error at /usr/lib/llvm-16/lib/cmake/clang/ClangTargets.cmake:792 (message):
      The imported target "clangBasic" references the file

         "/usr/lib/llvm-16/lib/libclangBasic.a"

      but this file does not exist.  Possible reasons include:

      * The file was deleted, renamed, or moved to another location.

      * An install or uninstall procedure did not complete successfully.

      * The installation package was faulty and contained

         "/usr/lib/llvm-16/lib/cmake/clang/ClangTargets.cmake"

      but not all the files it references.

    Call Stack (most recent call first):
      /usr/lib/llvm-16/lib/cmake/clang/ClangConfig.cmake:20 (include)
      generator/CMakeLists.txt:7 (Find_Package)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
docker: remove symlink for libclang (not required anymore)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
docker: fix path to clang builtin headers for woboq code browser

By default woboq uses $version.$minor.$patch, while clang simply uses
$version:

    CMake Error at generator/CMakeLists.txt:77 (message):
      Could not find any clang builtins headers in
      /usr/lib/llvm-16/lib/clang/16.0.4/include

    # ls -d /usr/lib/llvm-16/lib/clang/16/include
    /usr/lib/llvm-16/lib/clang/16/include

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Use __powerpc64__ over __ppc64__ (clang 16 stops declaring __ppc64__)

From the release notes [1]:

    To match GCC, __ppc64__ is no longer defined on PowerPC64 targets. Use __powerpc64__ instead.

  [1]: https://releases.llvm.org/16.0.0/tools/clang/docs/ReleaseNotes.html

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Remove removed in clang 16 -fexperimental-new-pass-manager

"The -fexperimental-new-pass-manager and -fno-legacy-pass-manager flags
have been removed. These flags have been silently ignored since Clang
15."

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

rschu1ze commented on docs/en/development/build.md:
````diff
@@ -47,8 +47,8 @@ GCC as a compiler is not supported
 ``` bash
-export CC=clang-15
-export CXX=clang++-15
+export CC=clang-16
````
Let's also adjust "docs/en/development/build-cross-osx.md"

rschu1ze commented on cmake/sanitize.cmake:
```diff
@@ -10,9 +10,7 @@ set (SAN_FLAGS "${SAN_FLAGS} -g -fno-omit-frame-pointer -DSANITIZER")
 
 if (SANITIZE)
     if (SANITIZE STREQUAL "address")
-        # LLVM-15 has a bug in Address Sanitizer, preventing the usage of 'sanitize-address-use-after-scope',
```
This breaks asan builds for users still building with Clang 15. Afaik, there was a policy of supporting the last two stable clang versions. But I wouldn't mind too much, we could also force >= clang 16, see `CLANG_MINIMUM_VERSION` in cmake/tools.cmake.

rschu1ze reviewed (commented):
Clang-tidy [build](https://s3.amazonaws.com/clickhouse-builds/PRs/49678/6e9985a94d88d49169a48551ba8910c2b373d0de/binary_tidy/build_log.log) is unfortunately broken because of [new checks](https://releases.llvm.org/16.0.0/tools/clang/tools/extra/docs/ReleaseNotes.html#new-checks). Disable the noisy ones, fix the feasible ones?

Other test failures are unrelated. Looks like the latest CI run exposed an [msan bug](https://s3.amazonaws.com/clickhouse-test-reports/49678/6e9985a94d88d49169a48551ba8910c2b373d0de/stress_test__msan_/stderr.log) in vectorscan.

rschu1ze reviewed (commented):
\+ minor: Processors/Formats/Impl/MsgPackRowInputFormat.cpp has some special treatment for clang 15. Maybe check if that's obsolete? (of course assuming that clang-16 will be mandatory)

azat commented on src/Common/atomicRename.cpp:
```diff
@@ -42,7 +42,7 @@ namespace ErrorCodes
     #elif defined(__aarch64__)
         #define __NR_renameat2 276
-    #elif defined(__ppc64__)
+    #elif defined(__powerpc64__)
```
Later is not supported by clang anymore - [`a794177` (#49678)](https://github.com/ClickHouse/ClickHouse/pull/49678/commits/a794177dc3c018530aa5feb86a2070e55e86a41b)

azat committed:
Update instructions for building under OSX to clang-16

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Get back -fno-sanitize-address-use-after-scope quirk for llvm/clang-15

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Revert "Suppress clang-analyzer-cplusplus.NewDelete in MsgPackRowInputFormat"

Let's try to revert this quirk during upgrading to clang 16.

This reverts commit c1e70169d2eec449012ac1bb555347a840c9c181.

azat commented:
>+ minor: Processors/Formats/Impl/MsgPackRowInputFormat.cpp has some special treatment for clang 15. Maybe check if that's obsolete? (of course assuming that clang-16 will be mandatory)

Yes, good idea, let's try, this is actually the problem only for `clang-tidy`, doubt that someone will run clang-tidy on their own (well it is also the case for clangd, but let's try revert it anyway)

azat committed:
Merge remote-tracking branch 'u/master' into build/llvm-16

Conflicts:
	docker/test/codebrowser/Dockerfile

azat committed:
Disable misc-use-anonymous-namespace check in clang-tidy

It is too noisy during upgrade:

$ grep error: build_log.log | awk '{print $NF}' | sort | uniq -c
      1 [bugprone-standalone-empty,-warnings-as-errors]
      1 [misc-misplaced-const,-warnings-as-errors]
   2147 [misc-use-anonymous-namespace,-warnings-as-errors]
      6 [modernize-loop-convert,-warnings-as-errors]
      9 [readability-redundant-string-cstr,-warnings-as-errors]

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Suppress bugprone-standalone-empty in DatabaseMySQL

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Update 02417_opentelemetry_insert_on_distributed_table for clang 16

CI: https://s3.amazonaws.com/clickhouse-test-reports/49678/6e9985a94d88d49169a48551ba8910c2b373d0de/stateless_tests__debug__[1_5].html
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix modernize-loop-convert clang-tidy warning

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix readability-redundant-string-cstr clang-tidy warning

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix misc-misplaced-const clang-tidy warning

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix stack-use-after-scope in gtest_resource_manager_static

Reordering of the static variables should be enough to fix this:

    [ RUN      ] IOResourceStaticResourceManager.Prioritization
    =================================================================
    ==13==ERROR: AddressSanitizer: stack-use-after-scope on address 0x7f7d8d621970 at pc 0x5636b80dcbb2 bp 0x7f7c48e47dd0 sp 0x7f7c48e47dc8
    READ of size 8 at 0x7f7d8d621970 thread T3975 (ThreadPool)
        #0 0x5636b80dcbb1 in IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_1::operator()(long) const build_docker/./src/IO/Resource/tests/gtest_resource_manager_static.cpp:78:13
        #1 0x5636b80dcbb1 in IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0::operator()() const build_docker/./src/IO/Resource/tests/gtest_resource_manager_static.cpp:95:21
        #2 0x5636b80dcbb1 in decltype(std::declval<IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0&>()()) std::__1::__invoke[abi:v15000]<IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0&>(IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23
        #3 0x5636b80dcbb1 in decltype(auto) std::__1::__apply_tuple_impl[abi:v15000]<IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0&, std::__1::tuple<>&>(IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0&, std::__1::tuple<>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/tuple:1789:1
        #4 0x5636b80dcbb1 in decltype(auto) std::__1::apply[abi:v15000]<IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0&, std::__1::tuple<>&>(IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0&, std::__1::tuple<>&) build_docker/./contrib/llvm-project/libcxx/include/tuple:1798:1
        #5 0x5636b80dcbb1 in ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0>(IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0&&)::'lambda'()::operator()() build_docker/./src/Common/ThreadPool.h:228:13
        #6 0x5636b80dcbb1 in decltype(std::declval<IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0>()()) std::__1::__invoke[abi:v15000]<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0>(IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0&&)::'lambda'()&>(IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23
        #7 0x5636b80dcbb1 in void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0>(IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0&&)::'lambda'()&>(ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0>(IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0&&)::'lambda'()&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9
        #8 0x5636b80dcbb1 in std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0>(IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0&&)::'lambda'(), void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12
        #9 0x5636b80dcbb1 in void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0>(IOResourceStaticResourceManager_Prioritization_Test::TestBody()::$_0&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16
        #10 0x5636ea219310 in std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16
        #11 0x5636ea219310 in std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12
        #12 0x5636ea219310 in ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) build_docker/./src/Common/ThreadPool.cpp:416:13
        #13 0x5636ea2258ac in void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:180:73
        #14 0x5636ea2258ac in decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23
        #15 0x5636ea2258ac in void std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/thread:284:5
        #16 0x5636ea2258ac in void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) build_docker/./contrib/llvm-project/libcxx/include/thread:295:5
        #17 0x7f7d8fcf1608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x8608) (BuildId: 7b4536f41cdaa5888408e82d0836e33dcf436466)
        #18 0x7f7d8fc16132 in __clone (/lib/x86_64-linux-gnu/libc.so.6+0x11f132) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)

    Address 0x7f7d8d621970 is located in stack of thread T0 at offset 368 in frame
        #0 0x5636b80d9bef in IOResourceStaticResourceManager_Prioritization_Test::TestBody() build_docker/./src/IO/Resource/tests/gtest_resource_manager_static.cpp:46

      This frame has 10 object(s):
        [32, 36) 'requests_per_thead' (line 48)
        [48, 208) 't' (line 49)
        [272, 296) 'ref.tmp' (line 51)
        [336, 352) 'last_priority' (line 74)
        [368, 376) 'check' (line 75) <== Memory access at offset 368 is inside this variable
        [400, 424) 'name' (line 83)
        [464, 512) 'ref.tmp13' (line 87)
        [544, 560) 'c' (line 101)
        [576, 600) 'ref.tmp32' (line 101)
        [640, 664) 'ref.tmp41' (line 102)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
There are few issues left, need to take a closer look are those real or not:

>[Integration tests (asan) [6/6] — fail: 14, passed: 424, flaky: 0](https://s3.amazonaws.com/clickhouse-test-reports/49678/6e9985a94d88d49169a48551ba8910c2b373d0de/integration_tests__asan__[6_6].html)

```
==1==ERROR: AddressSanitizer: stack-use-after-scope on address 0x7f7540186b20 at pc 0x55656ef458ad bp 0x7f754114acf0 sp 0x7f754114a4b8
READ of size 2 at 0x7f7540186b20 thread T2 (TCPHandler)
    #0 0x55656ef458ac in __asan_memcpy (/usr/bin/clickhouse+0x11b028ac) (BuildId: 6ba63057df561f6dc16fb8fd7623fd858bf327d1)
    #1 0x55659f62daf8 in rd_kafka_msg_new00 build_docker/./contrib/librdkafka/src/rdkafka_msg.c:192:3
    #2 0x55659f62daf8 in rd_kafka_msg_new0 build_docker/./contrib/librdkafka/src/rdkafka_msg.c:257:8
    #3 0x55659f630635 in rd_kafka_producev build_docker/./contrib/librdkafka/src/rdkafka_msg.c:642:23
    #4 0x55659f4a865c in cppkafka::Producer::do_produce(cppkafka::MessageBuilder const&, cppkafka::HeaderList<cppkafka::Header<cppkafka::Buffer>>&&) build_docker/./contrib/cppkafka/src/producer.cpp:128:19
    #5 0x55659f4a8306 in cppkafka::Producer::produce(cppkafka::MessageBuilder const&) build_docker/./contrib/cppkafka/src/producer.cpp:69:5
    #6 0x5565961e748c in DB::KafkaProducer::produce(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, unsigned long, std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::imm
    #7 0x5565962155b2 in DB::MessageQueueSink::consume(DB::Chunk) build_docker/./src/Storages/MessageQueueSink.cpp:64:23
    #8 0x5565980b9e8e in DB::SinkToStorage::onConsume(DB::Chunk) build_docker/./src/Processors/Sinks/SinkToStorage.cpp:18:5
    #9 0x556597e8a7eb in DB::ExceptionKeepingTransform::work()::$_1::operator()() const build_docker/./src/Processors/Transforms/ExceptionKeepingTransform.cpp:151:51
    #10 0x556597e8a7eb in decltype(std::declval<DB::ExceptionKeepingTransform::work()::$_1&>()()) std::__1::__invoke[abi:v15000]<DB::ExceptionKeepingTransform::work()::$_1&>(DB::ExceptionKeepingTransform::work()::$_1&) build_docker/./contrib/llvm-project/libcxx/in
    #11 0x556597e8a7eb in void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::ExceptionKeepingTransform::work()::$_1&>(DB::ExceptionKeepingTransform::work()::$_1&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9
    #12 0x556597e8a7eb in std::__1::__function::__default_alloc_func<DB::ExceptionKeepingTransform::work()::$_1, void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12
    #13 0x556597e8a7eb in void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::ExceptionKeepingTransform::work()::$_1, void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-
    #14 0x556597e8a2db in std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16
    #15 0x556597e8a2db in std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12
    #16 0x556597e8a2db in DB::runStep(std::__1::function<void ()>, DB::ThreadStatus*, std::__1::atomic<unsigned long>*) build_docker/./src/Processors/Transforms/ExceptionKeepingTransform.cpp:115:9
    #17 0x556597e88fd6 in DB::ExceptionKeepingTransform::work() build_docker/./src/Processors/Transforms/ExceptionKeepingTransform.cpp:151:34
    #18 0x5565977b6373 in DB::executeJob(DB::ExecutingGraph::Node*, DB::ReadProgressCallback*) build_docker/./src/Processors/Executors/ExecutionThreadContext.cpp:47:26
    #19 0x5565977b6373 in DB::ExecutionThreadContext::executeTask() build_docker/./src/Processors/Executors/ExecutionThreadContext.cpp:92:9
    #20 0x556597798c84 in DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) build_docker/./src/Processors/Executors/PipelineExecutor.cpp:255:26
    #21 0x556597797d80 in DB::PipelineExecutor::executeStep(std::__1::atomic<bool>*) build_docker/./src/Processors/Executors/PipelineExecutor.cpp:151:5
    #22 0x5565976a6756 in auto DB::TCPHandler::processInsertQuery()::$_0::operator()<DB::PushingPipelineExecutor>(DB::PushingPipelineExecutor&) const build_docker/./src/Server/TCPHandler.cpp:775:22
    #23 0x5565976a6756 in DB::TCPHandler::processInsertQuery() build_docker/./src/Server/TCPHandler.cpp:786:9
    #24 0x55659768f95a in DB::TCPHandler::runImpl() build_docker/./src/Server/TCPHandler.cpp:437:17
    #25 0x5565976c4b37 in DB::TCPHandler::run() build_docker/./src/Server/TCPHandler.cpp:2043:9
    #26 0x55659fbf0c2e in Poco::Net::TCPServerConnection::start() build_docker/./base/poco/Net/src/TCPServerConnection.cpp:43:3
    #27 0x55659fbf19f6 in Poco::Net::TCPServerDispatcher::run() build_docker/./base/poco/Net/src/TCPServerDispatcher.cpp:115:20
    #28 0x5565a00c143b in Poco::PooledThread::run() build_docker/./base/poco/Foundation/src/ThreadPool.cpp:188:14
    #29 0x5565a00ba4e7 in Poco::ThreadImpl::runnableEntry(void*) build_docker/./base/poco/Foundation/src/Thread_POSIX.cpp:335:27
    #30 0x7f768b994608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x8608) (BuildId: 7b4536f41cdaa5888408e82d0836e33dcf436466)
    #31 0x7f768b8b9132 in __clone (/lib/x86_64-linux-gnu/libc.so.6+0x11f132) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)

Address 0x7f7540186b20 is located in stack of thread T2 (TCPHandler) at offset 288 in frame
    #0 0x5565961e6bef in DB::KafkaProducer::produce(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, unsigned long, std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::imm

  This frame has 7 object(s):
    [32, 152) 'builder' (line 48)
    [192, 216) 'agg.tmp2'
    [256, 272) 'ref.tmp' (line 49)
    [288, 312) 'key_data' (line 55) <== Memory access at offset 288 is inside this variable
    [352, 368) 'ref.tmp20' (line 56)
    [384, 388) 'ref.tmp54' (line 74)
    [400, 404) 'ref.tmp59' (line 74)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
```

>[Stateful tests (msan) — Tests are not finished, fail: 1, passed: 55, skipped: 1](https://s3.amazonaws.com/clickhouse-test-reports/49678/6e9985a94d88d49169a48551ba8910c2b373d0de/stateful_tests__msan_.html)

```
==804==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x55eaf89711f5 in roseCheckMultipathShufti64 program_runtime.c
    #1 0x55eaf8964bb1 in roseRunProgram program_runtime.c
    #2 0x55eaf8956c74 in roseFloatingCallback match.c
    #3 0x55eaf8071089 in fdr_exec_teddy_msks3_pck teddy.c
    #4 0x55eaf8013490 in fdrExec fdr.c
    #5 0x55eaf80122f9 in hwlmExec hwlm.c
    #6 0x55eaf891c8d8 in roseBlockExec block.c
    #7 0x55eaf7fddd1c in hs_scan (/usr/bin/clickhouse+0x33c2cd1c) (BuildId: 322f1ef5851d4beda658c816a3b088ba278bc849)
    #8 0x55eae39a53eb in DB::FunctionsMultiStringFuzzySearch<DB::MultiMatchAnyImpl<DB::(anonymous namespace)::NameMultiFuzzyMatchAny, char8_t, (DB::MultiMatchTraits::Find)0, true>>::executeImpl(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName>> const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long) const multiFuzzyMatchAny.cpp
    #9 0x55ead361d2b7 in DB::FunctionToExecutableFunctionAdaptor::executeImpl(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName>> const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long) const (/usr/bin/clickhouse+0xf26c2b7) (BuildId: 322f1ef5851d4beda658c816a3b088ba278bc849)
    #10 0x55eafe85a6fc in DB::IExecutableFunction::executeWithoutLowCardinalityColumns(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName>> const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const build_docker/./src/Functions/IFunction.cpp:247:15
    #11 0x55eafe85c604 in DB::IExecutableFunction::executeWithoutSparseColumns(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName>> const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const build_docker/./src/Functions/IFunction.cpp:302:22
    #12 0x55eafe862377 in DB::IExecutableFunction::execute(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName>> const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const build_docker/./src/Functions/IFunction.cpp:374:16
    #13 0x55eb00a73ac0 in DB::executeAction(DB::ExpressionActions::Action const&, DB::(anonymous namespace)::ExecutionContext&, bool) build_docker/./src/Interpreters/ExpressionActions.cpp:613:60
    #14 0x55eb00a73ac0 in DB::ExpressionActions::execute(DB::Block&, unsigned long&, bool) const build_docker/./src/Interpreters/ExpressionActions.cpp:730:13
    #15 0x55eb07c879dd in DB::FilterTransform::doTransform(DB::Chunk&) build_docker/./src/Processors/Transforms/FilterTransform.cpp:135:25
    #16 0x55eb07c87503 in DB::FilterTransform::transform(DB::Chunk&) build_docker/./src/Processors/Transforms/FilterTransform.cpp:120:5
    #17 0x55eaf48b1940 in DB::ISimpleTransform::transform(DB::Chunk&, DB::Chunk&) build_docker/./src/Processors/ISimpleTransform.h:32:9
    #18 0x55eb073b2d3d in DB::ISimpleTransform::work() build_docker/./src/Processors/ISimpleTransform.cpp:89:9
    #19 0x55eb0741662d in DB::executeJob(DB::ExecutingGraph::Node*, DB::ReadProgressCallback*) build_docker/./src/Processors/Executors/ExecutionThreadContext.cpp:47:26
    #20 0x55eb0741662d in DB::ExecutionThreadContext::executeTask() build_docker/./src/Processors/Executors/ExecutionThreadContext.cpp:92:9
    #21 0x55eb073ea09c in DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) build_docker/./src/Processors/Executors/PipelineExecutor.cpp:255:26
    #22 0x55eb073ed155 in DB::PipelineExecutor::executeSingleThread(unsigned long) build_docker/./src/Processors/Executors/PipelineExecutor.cpp:221:5
    #23 0x55eb073ed155 in DB::PipelineExecutor::spawnThreads()::$_0::operator()() const build_docker/./src/Processors/Executors/PipelineExecutor.cpp:343:17
    #24 0x55eb073ed155 in decltype(std::declval<DB::PipelineExecutor::spawnThreads()::$_0&>()()) std::__1::__invoke[abi:v15000]<DB::PipelineExecutor::spawnThreads()::$_0&>(DB::PipelineExecutor::spawnThreads()::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23
    #25 0x55eb073ed155 in void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::PipelineExecutor::spawnThreads()::$_0&>(DB::PipelineExecutor::spawnThreads()::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9
    #26 0x55eb073ed155 in std::__1::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12
    #27 0x55eb073ed155 in void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16
    #28 0x55eaeaacf290 in std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16
    #29 0x55eaeaacf290 in std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12
    #30 0x55eaeaacf290 in ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__1::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) build_docker/./src/Common/ThreadPool.cpp:416:13
    #31 0x55eaeaada60a in void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:180:73
    #32 0x55eaeaada60a in decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23
    #33 0x55eaeaada60a in decltype(auto) std::__1::__apply_tuple_impl[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()&, std::__1::tuple<>&>(void&&, std::__1::tuple<>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/tuple:1789:1
    #34 0x55eaeaada60a in decltype(auto) std::__1::apply[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()&, std::__1::tuple<>&>(void&&, std::__1::tuple<>&) build_docker/./contrib/llvm-project/libcxx/include/tuple:1798:1
    #35 0x55eaeaada60a in ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()::operator()() build_docker/./src/Common/ThreadPool.h:228:13
    #36 0x55eaeaada4be in decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23
    #37 0x55eaeaada4be in void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9
    #38 0x55eaeaada4be in std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12
    #39 0x55eaeaada4be in void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16
    #40 0x55eaeaac651e in std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16
    #41 0x55eaeaac651e in std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12
    #42 0x55eaeaac651e in ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) build_docker/./src/Common/ThreadPool.cpp:416:13
    #43 0x55eaeaad3a6a in void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:180:73
    #44 0x55eaeaad3a6a in decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23
    #45 0x55eaeaad3a6a in void std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/thread:284:5
    #46 0x55eaeaad3a6a in void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) build_docker/./contrib/llvm-project/libcxx/include/thread:295:5
    #47 0x7f385456b608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x8608) (BuildId: 7b4536f41cdaa5888408e82d0836e33dcf436466)
    #48 0x7f3854490132 in __clone (/lib/x86_64-linux-gnu/libc.so.6+0x11f132) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)

  Uninitialized value was created by an allocation of 'data.i' in the stack frame
    #0 0x55eaf896efca in roseCheckMultipathShufti64 program_runtime.c
```

itrofimow commented:
Chiming in on stack-use-after-scope:
From a quick glance seems to be a valid one:
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/Kafka/KafkaProducer.cpp#L56
https://github.com/mfontanini/cppkafka/blob/master/include/cppkafka/buffer.h#L46

azat committed:
Fix 02417_opentelemetry_insert_on_distributed_table (wrong sed was used)

azat committed:
Fix stack-use-after-scope in KafkaProducer::produce()

ASan report:

    ==1==ERROR: AddressSanitizer: stack-use-after-scope on address 0x7f7540186b20 at pc 0x55656ef458ad bp 0x7f754114acf0 sp 0x7f754114a4b8
    READ of size 2 at 0x7f7540186b20 thread T2 (TCPHandler)
        0 0x55656ef458ac in __asan_memcpy (/usr/bin/clickhouse+0x11b028ac) (BuildId: 6ba63057df561f6dc16fb8fd7623fd858bf327d1)
        1 0x55659f62daf8 in rd_kafka_msg_new00 build_docker/./contrib/librdkafka/src/rdkafka_msg.c:192:3
        2 0x55659f62daf8 in rd_kafka_msg_new0 build_docker/./contrib/librdkafka/src/rdkafka_msg.c:257:8
        3 0x55659f630635 in rd_kafka_producev build_docker/./contrib/librdkafka/src/rdkafka_msg.c:642:23
        4 0x55659f4a865c in cppkafka::Producer::do_produce() build_docker/./contrib/cppkafka/src/producer.cpp:128:19
        5 0x55659f4a8306 in cppkafka::Producer::produce() build_docker/./contrib/cppkafka/src/producer.cpp:69:5
        6 0x5565961e748c in DB::KafkaProducer::produce() build_docker/./src/Storages/Kafka/KafkaProducer.cpp:70:23
        7 0x5565962155b2 in DB::MessageQueueSink::consume(DB::Chunk) build_docker/./src/Storages/MessageQueueSink.cpp:64:23
        8 0x5565980b9e8e in DB::SinkToStorage::onConsume(DB::Chunk) build_docker/./src/Processors/Sinks/SinkToStorage.cpp:18:5

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Update vectorscan library with fixes for MSan

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Merge remote-tracking branch 'u/master' into build/llvm-16

azat committed:
Update libfiu to fix the build

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>DocsCheck / CheckLabels (pull_request) — CheckLabels

@Felixoid maybe you can help? (Or someone else who looking at this comment)
One of this checks failed 4 times in a row, and stops the CI, any clue? Just a spot instances thing? (Although I've never seen that it will fail that frequently)
Also maybe they can be made retriable also?

azat committed:
Merge branch 'master' into build/llvm-16

alexey-milovidov committed:
Merge branch 'master' into build/llvm-16

azat committed:
Update 02417_opentelemetry_insert_on_distributed_table reference one more time (sigh)

This time, checked locally

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

rschu1ze committed:
Fix MsgPackRowInputFormat.cpp build

rschu1ze commented:
The local clang-tidy build gives me this error:

```
/home/ubuntu/repo/ch5/base/base/preciseExp10.cpp:41:25: error: 'p10' is an unsafe buffer that does not perform bounds checks [clang-diagnostic-unsafe-buffer-usage]
    static const double p10[]                                                                                                                                      
    ~~~~~~~~~~~~~~~~~~~~^~~~~                                                                                                                                      
/home/ubuntu/repo/ch5/base/base/preciseExp10.cpp:86:59: note: used in buffer access here                                                                           
    return exp2(3.32192809488736234787031942948939 * y) * p10[static_cast<int>(n) + 323];                                                                          
                                                          ^~~                                                                                                      
3246 warnings and 1 error generated.                                                                                                                               
```

It doesn't happen in normal builds. Repro:
``` bash
cmake -B build -S . -DENABLE_CLANG_TIDY=1
cd build
ninja base/base/CMakeFiles/common.dir/preciseExp10.cpp.o
```

Weird enough, that warning is already disabled in "cmake/warnings.cmake" ...

azat commented:
>Weird enough, that warning is already disabled in "cmake/warnings.cmake" ...

Those warnings should be suppressed in the `.clang-tidy`, not `cmake/warnings.cmake`, right?
Which version of clang-tidy do you have locally?

rschu1ze commented:
Ah, I forget to pass `-DCMAKE_C_COMPILER=clang-16 -DCMAKE_CXX_COMPILER=clang++-16` to cmake ... (and the CI clang-tidy build is green anyways)

azat committed:
Suppress MSan warning in NgramDistanceImpl::unrollLowering()

NgramDistanceImpl::unrollLowering() relies on the fact that PODArray has
padding and it is OK to access more items.

Here is an MSan report:

    ==656==WARNING: MemorySanitizer: use-of-uninitialized-value
        0 0x557fd825485f in DB::NgramDistanceImpl<4ul, char8_t, false, true, false>::vectorConstant(DB::PODArray<char8_t, 4096ul, Allocator<false, false>, 63ul, 64ul> const&, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 63ul, 64ul> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, DB::PODArray<float, 4096ul, Allocator<false, false>, 63ul, 64ul>&) (/usr/bin/clickhouse+0x124d885f) (BuildId: 76773125d8739591c75d4f4d263a2ffe7ca96855)
        1 0x557fd824eb83 in DB::FunctionsStringSimilarity<DB::NgramDistanceImpl<4ul, char8_t, false, true, false>, DB::NameNgramSearchCaseInsensitive>::executeImpl(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName>> const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long) const (/usr/bin/clickhouse+0x124d2b83) (BuildId: 76773125d8739591c75d4f4d263a2ffe7ca96855)
        2 0x557fd50023b7 in DB::FunctionToExecutableFunctionAdaptor::executeImpl() const (/usr/bin/clickhouse+0xf2863b7) (BuildId: 76773125d8739591c75d4f4d263a2ffe7ca96855)

      Uninitialized value was stored to memory at
        0 0x557fd4f8da5a in __msan_memcpy (/usr/bin/clickhouse+0xf211a5a) (BuildId: 76773125d8739591c75d4f4d263a2ffe7ca96855)
        1 0x557fd8253803 in DB::NgramDistanceImpl<4ul, char8_t, false, true, false>::vectorConstant(DB::PODArray<char8_t, 4096ul, Allocator<false, false>, 63ul, 64ul> const&, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 63ul, 64ul> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, DB::PODArray<float, 4096ul, Allocator<false, false>, 63ul, 64ul>&) (/usr/bin/clickhouse+0x124d7803) (BuildId: 76773125d8739591c75d4f4d263a2ffe7ca96855)
        2 0x557fd824eb83 in DB::FunctionsStringSimilarity<DB::NgramDistanceImpl<4ul, char8_t, false, true, false>, DB::NameNgramSearchCaseInsensitive>::executeImpl(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName>> const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long) const (/usr/bin/clickhouse+0x124d2b83) (BuildId: 76773125d8739591c75d4f4d263a2ffe7ca96855)
        3 0x557fd50023b7 in DB::FunctionToExecutableFunctionAdaptor::executeImpl() const (/usr/bin/clickhouse+0xf2863b7) (BuildId: 76773125d8739591c75d4f4d263a2ffe7ca96855)

      Uninitialized value was stored to memory at
        0 0x557fd4f8da5a in __msan_memcpy (/usr/bin/clickhouse+0xf211a5a) (BuildId: 76773125d8739591c75d4f4d263a2ffe7ca96855)
        1 0x5580061699f5 in detail::memcpySmallAllowReadWriteOverflow15Impl(char*, char const*, long) build_docker/./src/Common/memcpySmall.h:42:13
        2 0x5580061699f5 in memcpySmallAllowReadWriteOverflow15(void*, void const*, unsigned long) build_docker/./src/Common/memcpySmall.h:57:5
        3 0x5580061699f5 in DB::ColumnString::replicate(DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 63ul, 64ul> const&) const build_docker/./src/Columns/ColumnString.cpp:462:13
        4 0x558005d3fae4 in DB::ColumnConst::convertToFullColumn() const build_docker/./src/Columns/ColumnConst.cpp:48:18

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Merge remote-tracking branch 'u/master' into build/llvm-16

* u/master:
  enable used flags's reinit only when the hash talbe rehash
  Fix build of libfiu on clang-16
  fix flaky test 02504_regexp_dictionary_ua_parser
  fix convertation
  Fix test that expected CH to apply a wrong optimization
  ActionsDAG: remove wrong optimization

azat commented:
Test failures looks unrelated:

>[Integration tests (release) [2/4] — fail: 2, passed: 545, flaky: 0](https://s3.amazonaws.com/clickhouse-test-reports/49678/e5c4eb35e8e85f948a44730a8970e40eeb4b7f6f/integration_tests__release__[2_4].html)

```
E           Code: 117. DB::Exception: Received from 172.16.1.7:9000. DB::Exception: Data directory for table already containing data parts - probably it was unclean DROP table or manual intervention. You must either clear directory by hand or use ATTACH TABLE instead of CREATE TABLE if you need to use that parts.. Stack trace:
```

>[Stateless tests (release, analyzer) — fail: 1, passed: 5246, skipped: 17](https://s3.amazonaws.com/clickhouse-test-reports/49678/e5c4eb35e8e85f948a44730a8970e40eeb4b7f6f/stateless_tests__release__analyzer_.html)

It shows bunch of `NOT_FAILED`, but apparently it simply does not count it as an error (since in upstream it has the same - https://s3.amazonaws.com/clickhouse-test-reports/0/fc993f4c07ae3d14e5141d810a22322d0f444949/stateless_tests__release__analyzer_.html and check is green)

But, there is one failure of `00992_system_parts_race_condition_zookeeper_long` while this test is flaky - [play](https://play.clickhouse.com/play?user=play#U0VMRUNUIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSBjaGVja19uYW1lIExJS0UgJ1N0YXRlJScKICAgIEFORCBjaGVja19zdGFydF90aW1lID49IG5vdygpIC0gSU5URVJWQUwgMSBEQVkKLy8gICAgQU5EIHB1bGxfcmVxdWVzdF9udW1iZXIgPSAwCi8vICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnU0tJUFBFRCcKICAgIEFORCB0ZXN0X3N0YXR1cyBMSUtFICdGJScKLy8gICAgQU5EIGNoZWNrX3N0YXR1cyAhPSAnc3VjY2VzcycKICAgIGFuZCB0ZXN0X25hbWUgbGlrZSAnJTAwOTkyX3N5c3RlbV9wYXJ0c19yYWNlX2NvbmRpdGlvbl96b29rZWVwZXJfbG9uZyUnCk9SREVSIEJZIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgY2hlY2tfc3RhcnRfdGltZQ==)

rschu1ze commented on .gitmodules:
```diff
@@ -267,7 +267,10 @@
-	url = https://github.com/VectorCamp/vectorscan
+	# FIXME: update once upstream fixes will be merged:
+	# - https://github.com/VectorCamp/vectorscan/pull/148
+	# - https://github.com/VectorCamp/vectorscan/pull/149
```
Just an interesting observation: "cmake/sanitize.cmake" sets `-fsanitize-blacklist=${CMAKE_SOURCE_DIR}/tests/msan_suppressions.txt` for msan. This file contains line `fun:roseRunProgram` that suppresses msan warnings in top-level (publicly callable) function "roseRunProgram ()" in rose/program_runtime.c. In other words, it achieves theoretically the same thing as https://github.com/VectorCamp/vectorscan/pull/149.

But the GCC devs agree with your approach, https://gcc.gnu.org/bugzilla/show_bug.cgi?id=61978#c1 :smile:.

Anyways, once this PR is in, I can try to clean out msan_suppressions.txt, it seems redundant and un-obvious to have an alternative way to suppress msan warnings.

azat replied:
>Just an interesting observation: "cmake/sanitize.cmake" sets -fsanitize-blacklist=${CMAKE_SOURCE_DIR}/tests/msan_suppressions.txt for msan. This file contains line fun:roseRunProgram that suppresses msan warnings in top-level (publicly callable) function "roseRunProgram ()" in rose/program_runtime.c. In other words, it achieves theoretically the same thing as https://github.com/VectorCamp/vectorscan/pull/149.

Hm, I totally forgot to look at MSan suppressions, thanks! I want check it again now, need to get full stack trace to compare why it does not work in debug build, and works in release.

There is actually one more fix for this - https://github.com/ClickHouse/boost/pull/12, I tried to fix this in the `small_vector`, but since clang-16 it was not easy (I wrote about this in the vectorscan PR)

>Anyways, once this PR is in, I can try to clean out msan_suppressions.txt, it seems redundant and un-obvious to have an alternative way to suppress msan warnings

Will take a look, but first, I need to check why it didn't work (maybe there is something more...)

rschu1ze commented on src/Functions/FunctionsStringSimilarity.cpp:
```diff
@@ -108,6 +108,10 @@ struct NgramDistanceImpl
         if constexpr (case_insensitive)
         {
+#if defined(MEMORY_SANITIZER)
+            /// Due to PODArray padding accessing more elements should be OK
```
Speaking of tests/msan_suppressions.txt, it contains another suppression

```
# We apply std::tolower to uninitialized padding, but don't use the result, so        
# it is OK. Reproduce with "select ngramDistanceCaseInsensitive(materialize(''), '')" 
fun:tolower                                                                           
```

Looks like msan16 no longer picks these suppressions up.

Felixoid commented on .gitmodules:
```diff
@@ -267,7 +267,10 @@
+	# FIXME: update once upstream fixes will be merged:
+	# - https://github.com/VectorCamp/vectorscan/pull/148
+	# - https://github.com/VectorCamp/vectorscan/pull/149
+	url = https://github.com/azat-ch/vectorscan
```
Sorry for being a bit late for the party, but there's no need to change the url. It is available all over the forks

https://github.com/VectorCamp/vectorscan/commit/aaca65aa210ce3ec91bd2b249c4d59e55e80a869

rschu1ze replied:
I am on it ...

rschu1ze replied:
I also don't mind too much. I would say the unspoken rule is "reference upstream if possible (= there are no custom patches), otherwise reference a fork in the ClickHouse organization".

Anyways, after making some silly mess, I now
- created https://github.com/ClickHouse/vectorscan/
- pushed your two fixes to [branch ClickHouse/vectorscan/5.4.9](https://github.com/ClickHouse/vectorscan/commits/ClickHouse/vectorscan/5.4.9) in this forked repo
- and referenced that fork + commit from ClickHouse ([here](https://github.com/ClickHouse/ClickHouse/pull/49848)).

Your two pushes against vectorscan upstream remain as is (fingers crossed they are merged).

Felixoid commented:
Unfortunately, there's another leftover https://github.com/ClickHouse/ClickHouse/actions/runs/4964847326/jobs/8885617747#step:6:781

And the report with full logs https://s3.amazonaws.com/clickhouse-test-reports/0/6df457dcaf32317090fc611138b5bd1e1ff5b8ce/woboq_build.html

Here's the last success build with 15 https://s3.amazonaws.com/clickhouse-test-reports/49701/4dcdcf4c11b5604f1c5d3121c9c6fea3e957b605/woboq_build.html
